### ISSUE TITLE
wtmp record mangling could use some tuning

### ISSUE THREAD

**Issue by wesolows** [oxidecomputer/omicron#3514]
After #2516 we note have `boottime_rewrite` and co. to rewrite the boot time record in utmp/wtmp; see https://github.com/oxidecomputer/omicron/blob/c2de4805acfd0c20bef9d87afacac9f8e821ad87/sled-agent/src/services.rs#L2525-L2565.  The original intent of this was to make code that reads this record (notably but probably not exclusively uptime(1)) work "normally" to reduce surprises for post-mortem debugging with respect to the contents of support bundles, as well as anyone doing bringup or engineering work on Oxide-owned machines who might be able to execute commands.

As it stands today, this has two flaws:

1. It writes the *current* wall time as the boot time.  This is kind of almost right, but invariably sled-agent starts up -- and NTP completes synching the wall clock -- some seconds or more likely minutes after the kernel actually started keeping time.  This isn't catastrophic since at least the record is vaguely plausible, unlike 28 Dec 1986, but it's just wrong enough that it might confuse someone.  We should instead subtract from the current NTP-supplied wall time the amount of time the sled has been running.  This should allow us sub-second accuracy which is basically fine and probably as good as we get on other illumos systems anyway.  In particular, `CLOCK_MONOTONIC` on illumos is `CLOCK_HIGHRES` which should not be affected by leap seconds or NTP adjustments (whether slow or fast).  Both `gethrtime()` and `clock_gettime(CLOCK_HIGHRES)` should supply the absolute time since the kernel started its clock on illumos.  `CLOCK_MONOTONIC` is not guaranteed to do this by POSIX, and in particular it does not on Linux, but this code isn't intended to be platform-agnostic and if for some reason (test mocks?) it needed to be, there are nonportable alternatives for most other platforms we might care about.
2. It runs every time sled-agent starts up, which means that effectively we write a record that tells consumers the sled booted whenever sled-agent last started.  Normally, sled-agent starts relatively soon after boot and should never restart, but that's certainly not guaranteed: it can crash and recover, and in the environment where this is most important (non-production machines where control plane engineering is actively occurring) it's likely that sled-agent may be manually restarted for any number of reasons.  This should be done only once.  There are any number of ways to accomplish that within sled-agent, or one could examine the current boot time record to see whether it's plausibly correct and write a new record only if it is not.  One might imagine that "plausibly correct" is a hard thing to define, but it is not: no time before 9 Sep 2019 is ever plausible for anything done at Oxide, so all we have to do is guarantee that the kernel won't start its wall clock at a time later than that without reliable knowledge of the true wall time, a condition that's currently met and we have no reason to change.

Note that this probably doesn't apply to PCs, unless it's a PC with a broken RTC.  But it will always apply to Gimlet and almost certainly to future sleds as well.

**Comment by wesolows** [comment]
As an optional addition, we could set the kernel's `boot_time` variable.  This is intended to be the UTC wall time, in seconds, that corresponds to a `CLOCK_HIGHRES` value of 0, or at least approximately so.  There is currently no interface to perform this operation, so if we wanted to do anything better than `mdb -kw`, we'd need a companion OS ticket to provide one.  This piece could also be deferred; it might make post-mortem debugging of crash dumps slightly easier, but this variable isn't used by mdb or any dmods and other means are available (and employed) to determine the amount of time that has elapsed since timekeeping began.

**Comment by davepacheco** [comment]
See also https://github.com/jclulow/junk-gimlet-fixtime/blob/main/src/main.rs
